### PR TITLE
Add scrolling to support to grid

### LINKED ISSUE #2332: scrolling with mouse to pick and put items
#### Issue description:
want to pick up and restore items with mouse scroll (like regular chests)
#### What happens:
nothing
#### What you expected to happen:
want to pick up and restore items with mouse scroll (like regular chests)
#### Steps to reproduce:
just use scroll and nothing will happen :)))
#### Version (make sure you are on the latest version before reporting):

- Minecraft: dont know, playing on DDSS modpack
- Forge: dont know, playing on DDSS modpack
- Refined Storage: dont know, playing on DDSS modpack

Does this issue occur on a server? [yes/no] no

#### If a (crash)log is relevant for this issue, link it here:

[pastebin/gist/etc link here]

### PR DESCRIPTION
**Pull request by Darkere**
port of #2351 to 1.16
resolves #2332

### CODE REVIEW

File: src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java [hunk 1]
```diff
@@ -31,7 +32,14 @@ public ItemGridHandler(INetwork network) {
     }
 
     @Override
-    public void onExtract(ServerPlayerEntity player, UUID id, int flags) {
+    public void onExtract(ServerPlayerEntity player, ItemStack stack, int preferredSlot, int flags) {
+        if (network.getItemStorageCache().getList().getEntry(stack, IComparer.COMPARE_NBT) != null) {
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java:36]: Assign `network.getItemStorageCache().getList().getEntry(stack, IComparer.COMPARE_NBT)` to a variable

File: src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java [hunk 2]
```diff
@@ -82,11 +90,22 @@ public void onExtract(ServerPlayerEntity player, UUID id, int flags) {
         if (!took.isEmpty()) {
             if ((flags & EXTRACT_SHIFT) == EXTRACT_SHIFT) {
                 IItemHandler playerInventory = player.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, Direction.UP).orElse(null);
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java:92]: Can't we do a `ifPresent`, much cleaner.

File: src/main/java/com/refinedmods/refinedstorage/network/ServerProxy.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import com.refinedmods.refinedstorage.apiimpl.network.grid.handler.ItemGridHandler;
+import com.refinedmods.refinedstorage.apiimpl.storage.cache.ItemStorageCache;
+import com.refinedmods.refinedstorage.container.GridContainer;
+import net.minecraft.entity.player.ServerPlayerEntity;
+import net.minecraft.item.ItemStack;
+
+import java.util.UUID;
+
+public class ServerProxy {
+    public static void onInventoryScrollMessage(ServerPlayerEntity player, int slot, boolean shift, boolean up) {
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/network/ServerProxy.java:14]: This impl should go in the grid handler, perhaps also add a way to reuse this logic.

File: src/main/java/com/refinedmods/refinedstorage/network/ServerProxy.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                if (stackOnCursor.isEmpty()) {
+                    itemHandler.onExtract(player, stackInSlot, -1, flags);
+                } else {
+                    itemHandler.onExtract(player, stackOnCursor, -1, flags);
+                }
+            }
+        }
+    }
+
+    public static void onGridScrollMessage(ServerPlayerEntity player, UUID id, boolean shift, boolean ctrl, boolean up) {
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/network/ServerProxy.java:51]: This impl should go in the grid handler, perhaps also add a way to reuse this logic.

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 5]
```diff
@@ -86,7 +89,6 @@ public void onPageChanged(int page) {
     @Override
     protected void onPreInit() {
         super.onPreInit();
-        this.doSort = true;
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:89]: Why did you move this?

**Review by raoulvdberge — CHANGES_REQUESTED** [review]
Please also add changelog entries.

File: src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java (hunk 2 above)
Comment (Darkere) [reply src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java:92]: If I put it into a lambda I need to make size final and convert took into an array :/

changed to resolve which is a bit cleaner

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java (hunk 5 above)
Comment (Darkere) [reply src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:89]: Oh wow. Forgot this was in here. This was something I forgot to commit in the shiftpreventssorting PR. Totally unnecessary

**Comment by Darkere** [comment]
Not sure what you mean by making it reusable.  I could make a Util class that takes over the boolean logic but I don't think passing around runnables makes a lot of sense here.

File: src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java [hunk 6]
```diff
@@ -89,21 +95,21 @@ public void onExtract(ServerPlayerEntity player, UUID id, int preferredSlot, int
 
         if (!took.isEmpty()) {
             if ((flags & EXTRACT_SHIFT) == EXTRACT_SHIFT) {
-                IItemHandler playerInventory = player.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, Direction.UP).orElse(null);
-                if (playerInventory != null) {
+                Optional<IItemHandler> playerInventory = player.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, Direction.UP).resolve();
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java:98]: Why can't you do this?
```
player.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, Direction.UP).ifPresent(playerInventory -> {
   // your logic here
});
```

File: src/main/java/com/refinedmods/refinedstorage/api/network/grid/handler/IItemGridHandler.java [hunk 7]
```diff
@@ -81,4 +81,25 @@
...
+
+    /**
+     * Called when a player shift or ctrl scrolls in the player inventory
+     *
+     * @param player player that is scrolling
+     * @param slot   slot the mouse is hovering over
+     * @param shift  if true shift is pressed, if false ctrl is pressed
+     * @param up     whether the player is scrolling up or down
+     */
+    void onInventoryScrollMessage(ServerPlayerEntity player, int slot, boolean shift, boolean up);
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/api/network/grid/handler/IItemGridHandler.java:93]: Why did you add "message" to the method name? The grid handlers shouldn't know about "message" implementation details.

File: src/main/java/com/refinedmods/refinedstorage/api/network/grid/handler/IItemGridHandler.java [hunk 8]
```diff
@@ -81,4 +81,25 @@
...
+    /**
+     * Called when a player shift or ctrl scrolls in the Grid View
+     *
+     * @param player player that is scrolling
+     * @param id     UUID of the GridStack that the mouse is hovering over or UUID(0,0) if not over a stack
+     * @param shift  true if shift is pressed
+     * @param ctrl   true if ctrl is pressed
+     * @param up     whether the player is scrolling up or down
+     */
+    void onGridScrollMessage(ServerPlayerEntity player, UUID id, boolean shift, boolean ctrl, boolean up);
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/api/network/grid/handler/IItemGridHandler.java:104]: Why did you add "message" to the method name? The grid handlers shouldn't know about "message" implementation details.

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 9]
```diff
@@ -86,7 +88,7 @@ public void onPageChanged(int page) {
...
+        doSort = true;
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:91]: Revert back to `this.doSort`

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 10]
```diff
@@ -65,7 +68,6 @@
 
     public GridScreen(GridContainer container, IGrid grid, PlayerInventory inventory, ITextComponent title) {
         super(container, 227, 0, inventory, title);
-
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:68]: Add back newline.

**Review by raoulvdberge — CHANGES_REQUESTED** [review]

File: src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java (hunk 6 above)
Comment (Darkere) [reply src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java:98]: This is what that would look like. Notices the "finalSize" and the took array. Your choice ofc. 

```java
int finalSize = size;
player.getCapability(CapabilityItemHandler.ITEM_HANDLER_CAPABILITY, Direction.UP).ifPresent(playerInventory -> {
                    if (preferredSlot != -1) {
                        ItemStack remainder = playerInventory.insertItem(preferredSlot, took[0], true);
                        if (remainder.getCount() != took[0].getCount()) {
                            ItemStack inserted = network.extractItem(item, finalSize - remainder.getCount(), Action.PERFORM);
                            playerInventory.insertItem(preferredSlot, inserted, false);
                            took[0].setCount(remainder.getCount());
                        }
                    }
                    if (!took[0].isEmpty()) {
                        if (ItemHandlerHelper.insertItemStacked(playerInventory, took[0], true).isEmpty()) {
                            took[0] = network.extractItem(item, finalSize, Action.PERFORM);

                            ItemHandlerHelper.insertItemStacked(playerInventory, took[0], false);
                        }
                    }
                });
```

Comment (raoulvdberge) [reply src/main/java/com/refinedmods/refinedstorage/apiimpl/network/grid/handler/ItemGridHandler.java:98]: OK, fair point. Let's leave it as it, much cleaner.
